records/state/state.go

```diff
@@ -162,7 +162,8 @@ func statusIPs(st []Status, src func(*Status) []string) []string {
 	// https://github.com/apache/mesos/blob/0.24.0/src/slave/slave.cpp#L5226-L5238
 	ts, j := -1.0, -1
 	for i := range st {
-		if st[i].State == "TASK_RUNNING" && st[i].Timestamp > ts {
+		state := st[i].State
+		if (state == "TASK_RUNNING" || state == "TASK_STAGING") && st[i].Timestamp > ts {
```

Comment:
TASK_STAGING could be a while if it's depending on a slow docker pull (see
below for ambiguity w/ respect to STAGING). We have a 5m timeout by default
for k8s-mesos and that's still not enough for some users large docker
images. To some degree the framework might want to play a role in
determining advertisement. For example, in DiscoveryInfo perhaps a
framework could apply a label to a Port such as
"mesos-dns/advertise-always" if those service instances that are unhealthy
should be advertised to the cluster anyway.

For the "bootstrap-clustered-systems" use case there's often a different
peer port vs. client port; port labels could help distinguish advertising
requirements here.

The default advertisement scheme for service instance ports for mesos-dns
could could assume "advertise only for healthy tasks".

Also, some tasks transition from STAGING to RUNNING. Others go from STAGING
to STARTING to RUNNING. There's no way that mesos-dns is going to be able
to reliably determine the meaning of these transitions across multiple
frameworks. E.g. a task in STAGING could be in the middle of a docker-pull.
Or it could be trying to connect to its peers in the cluster. Totally
depends on the framework in question.


On Wed, Oct 7, 2015 at 12:23 PM, Tomás Senart <notifications@github.com>
wrote:

> In records/state/state.go
> <https://github.com/mesosphere/mesos-dns/pull/308#discussion_r41411945>:
>
> > @@ -162,7 +162,8 @@ func statusIPs(st []Status, src func(*Status) []string) []string {
> >  	// https://github.com/apache/mesos/blob/0.24.0/src/slave/slave.cpp#L5226-L5238
> >  	ts, j := -1.0, -1
> >  	for i := range st {
> > -		if st[i].State == "TASK_RUNNING" && st[i].Timestamp > ts {
> > +		state := st[i].State
> > +		if (state == "TASK_RUNNING" || state == "TASK_STAGING") && st[i].Timestamp > ts {
>
> Yeah, although Mesos-DNS can't guarantee the health of a record. How long
> do you think a task can be in staging state? What would be the implications
> of clients connecting to services in that state? Wouldn't they need to
> retry in case of failure anyway?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mesosphere/mesos-dns/pull/308/files#r41411945>.
>